sway/lock.c

```diff
@@ -0,0 +1,174 @@
+#define _POSIX_C_SOURCE 200809L
+#include "log.h"
+#include "sway/input/keyboard.h"
+#include "sway/input/seat.h"
+#include "sway/lock.h"
+#include "sway/output.h"
+#include "sway/server.h"
+
+#define PERMALOCK_CLIENT (struct wl_client *)(-1)
+
+struct sway_session_lock_surface {
+	struct sway_output *output;
+	struct wlr_surface *surface;
+	struct wl_listener map;
+	struct wl_listener destroy;
+};
+
+static void handle_surface_map(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, map);
+	sway_force_focus(surf->surface);
+	output_damage_whole(surf->output);
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, destroy);
+	wl_list_remove(&surf->map.link);
+	wl_list_remove(&surf->destroy.link);
+	output_damage_whole(surf->output);
+	free(surf);
+}
+
+static void handle_new_surface(struct wl_listener *listener, void *data)
+{
+	//struct sway_session_lock_manager *lock_state = server.session_lock->data;
+	struct wlr_session_lock_surface_v1 *lock_surface = data;
+	struct sway_session_lock_surface* surf = calloc(1, sizeof(*surf));
+	if (surf == NULL)
+		return;
+
+	sway_log(SWAY_DEBUG, "new lock layer surface");
+
+	struct sway_output *output = lock_surface->output->data;
+	wlr_session_lock_surface_v1_configure(lock_surface, output->width, output->height);
+
+	surf->surface = lock_surface->surface;
+	surf->output = output;
+	surf->map.notify = handle_surface_map;
+	wl_signal_add(&lock_surface->events.map, &surf->map);
+	surf->destroy.notify = handle_surface_destroy;
+	wl_signal_add(&lock_surface->events.destroy, &surf->destroy);
+}
+
+static void handle_unlock(struct wl_listener *listener, void *data)
+{
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+	struct sway_seat *seat;
```

Comment:
Style: this variable can be declared right before the `wl_list_for_each` invocation